### DISCUSSION TITLE
Integration with existing CI/CD

### DISCUSSION THREAD

**Discussion by bduff9** [planetscale/discussion#121] [FAQs]
Hello!  Just saw the presentation at NextJS conf and wanted to test PlanetScale out.  I love the idea so far and understand the ease of use for manual DB migrations (especially love the drift check and non-blocking aspects).  However, I had some questions on how we might incorporate this into our current deployment workflow.

Today, we use TypeORM to generate migrations locally, which are then verified and tests are written against them.  We then commit and push them to our remote repo, where a pipeline automatically runs on any push to a main branch (develop or main).  This pipeline applies any new migrations to develop, and then the tests automatically run against them.  If everything passes, the code is then deployed to the develop env.  The same happens when merging to main for our production env.

Since we are only supposed to go through the PlanetScale process to apply changes (i.e. cannot alter prod DB directly), what is the best way to keep this or a similar workflow?  My understanding is that I can script up some changes using the CLI, but it still seems that between the development branching (off of main) and the approvals needed to apply the changes, it will be much harder if not impossible to keep our existing automation.

I am all for changing how we do things if it's better, so just trying to understand how the PlanetScale team envisions this, as for us, if we deploy code before the DB changes, it is likely to cause app errors.  Also, possible to get errors if we apply the DB changes too early before the updated code is deployed.  So trying to understand the best way for us to accomplish where the DB changes are applied as the code is deployed to keep everything in sync, versus trying to do 2 parallel workflows (code deploy and PS branching) at the same time to avoid errors.

Thanks for any thoughts, loving the UI and DX so far!

**Comment by johnyerhot** [comment]
👋  @bduff9 

> Today, we use TypeORM to generate migrations locally, which are then verified and tests are written against them. We then commit and push them to our remote repo, where a pipeline automatically runs on any push to a main branch (develop or main). This pipeline applies any new migrations to develop, and then the tests automatically run against them. If everything passes, the code is then deployed to the develop env. The same happens when merging to main for our production env.

If I'm understanding correctly, the way that PlanetScale would work here is that when migrations are created and pushed to one of your code branches like develop, inside that pipeline you could create a new PlanetScale database branch for that pipeline run and run tests against it like normally. After everything passes the database branch is then merged into the the parent database and the branch can be removed.

I think a similar process would be what you're looking at for your other code branches as well. It's likely that I've missed something important in your process and it might make sense to have a conversation with someone on our end to fully talk it out. Feel free to hit up our [contact](https://planetscale.com/contact) page or reach out to [support](https://planetscale.com/support) and we can make sure to get you to the right place.

**Reply by ydennisy** [reply]
@johnyerhot I think this question is quite a common one, I am also struggling to figure out the best approach for this with the different model you guys provide.

Would you mind suggesting some best practices here, specifically:
- how should we deploy branch updates as part of a PR review and CI/CD process with GH actions?
- how should branches be made when testing both locally and as part of CI/CD?
- do you suggest branches or DBs for environments i.e prod, staging etc
- what is the process post a schema migration to "backfill" and update current DB records with the schema update?

Working with PS locally has been fine, and the docs cover most things, but how to connect up PS, with code review, automated tests, build process and production is not fully clear.

@bduff9 sorry for the hijack!

**Reply by johnyerhot** [reply]
Hi @ydennisy! 

You're right this is a quite common topic. The answers are going to be highly dependent on how current processes and needs are, as well as tons of other factors.  Internally we've been talking about the best way to tell the story as it varies very much between organizations.

That all said, I can talk a bit about how we work internally to use PlanetScale to develop PlanetScale. I am not a professional blog or technical writer but hopefully it's helpful until we get a bit more public-facing stuff out there.


> how should we deploy branch updates as part of a PR review and CI/CD process with GH actions?

Internally here any PR that has a database schema change associated also has a referenced DR (Deploy Request) in the Pull Request. Usually we put "depends on <link to DR>" in the main body of the Pull Request. They're reviewed in parallel and code updates and database updates are deployed in a similar way to what @bduff9 walked through [here](https://github.com/planetscale/beta/discussions/122#discussioncomment-1579439) so not to cause downtime. 

This may mean a new feature requires multiple pull requests to be merged in order (internally once something is merged to `main`, it's automatically deployed) with DRs deployed in-between to ensure everything works correctly.

> how should branches be made when testing both locally and as part of CI/CD?

The [`pscale`](https://github.com/planetscale/cli#installation) CLI is what you'll want to use in any automated environment to create branches, like in CI. It's also easier locally in my opinion. I am also a terminal junkie 😸 . 

In CI you'll generally want to create the branch and then seed data, if needed. 

> do you suggest branches or DBs for environments i.e prod, staging etc

This one is going to very much depend on your environment and development processes. Production, in general, is going to be your `main` branch. It will have certain protections and whatnot that you'll want for production. Staging may be better off as a branch, though. Same with testing.

There may be some cases where completely different staging/development/testing database may be the right thing, however.

> what is the process post a schema migration to "backfill" and update current DB records with the schema update?

Good question - how we and many folks do it is to make the schema change and then run a job that does the back-filling. Depending on what tools you're using this could be as simple as a bash script or a Rake Task in Ruby. 

After the job has completed and the data is validated, deploy the code to take advantage of the new schema and finally the schema can be safely modified to remove the old stuff.

A good way to test the job is to [backup](https://docs.planetscale.com/concepts/back-up-and-restore) your `main` database, restore it to another branch, and test it there.

Hope that's helpful!

**Reply by ydennisy** [reply]
Hi @johnyerhot 

Thanks for your reply.

> Staging may be better off as a branch, though. Same with testing.

Can you elaborate here, why is that your guidance?

> There may be some cases where completely different staging/development/testing database may be the right thing, however.

Again what are the cases?

I appreciate your replies, but when you are providing different paradigms in comparison to what is traditionally used, I feel you are required to explain in what situations they are useful. It is far too easy to say that is depends on each use case.

**Reply by ydennisy** [reply]
> Good question - how we and many folks do it is to make the schema change and then run a job that does the back-filling. Depending on what tools you're using this could be as simple as a bash script or a Rake Task in Ruby.
After the job has completed and the data is validated, deploy the code to take advantage of the new schema and finally the schema can be safely modified to remove the old stuff.

Again sorry but could you elaborate a little here? Lets imagine we make a schema change, and then want to trigger this all as part of a CI/CD process in something like GH actions, could you spell out step by step what you guys would do? Where does the data update run, on the CI/CD machine?

**Comment by johnyerhot** [comment]
👋 @ydennisy thanks for following up!

> >  Staging may be better off as a branch, though. Same with testing.
> Can you elaborate here, why is that your guidance?

I wouldn't say this is necessarily guidance or what we're saying you should do, however, it's the model we follow and can confidently say other folks use. We consider branches, both code and database, to be short lived and single purpose very similar to Git branches in the [GitHub Flow](https://docs.github.com/en/get-started/quickstart/github-flow).

>  >  There may be some cases where completely different staging/development/testing database may be the right thing, however.
> Again what are the cases?

What I was thinking of were cases where the weight of swapping out an existing workflow with testing, staging, and develop databases or code branches that are already permanent fixtures costs too much or the organization is too large to undertake. I'm sure there are other reasons I'm not thinking of either.

Another factor is that branches on PlanetScale have a cost. The number of open dev branches and concurrent CI runs would be limited by the number of PlanetScale branches you pay for. It may also just be plain faster to run against a local MySQL 8 install depending on where your developers or CI servers are located. 

> I appreciate your replies, but when you are providing different paradigms in comparison to what is traditionally used, I feel you are required to explain in what situations they are useful. It is far too easy to say that is depends on each use case.

I get that.

It's like [here is this database specific workflow](https://docs.planetscale.com/concepts/planetscale-workflow) and then the question is how do I use that in my current workflow that depends on databases? Right now "it depends" is the best I can do for you. I know that's now exactly the answer you're looking for, but it's the honest one for right now.

> Again sorry but could you elaborate a little here? Lets imagine we make a schema change, and then want to trigger this all as part of a CI/CD process in something like GH actions, could you spell out step by step what you guys would do? Where does the data update run, on the CI/CD machine?

Internally, we run anything that will be transforming data manually. So, in that case, we'd manually be running the script/job/whatever and vetting it against a non-production branch before running it in production.

If you're super interested in what might work best for y’all specifically I'm happy to put you in touch with some of our sales engineers who basically live for implementation detail discussions haha. Might be helpful 😸

**Comment by johnyerhot** [comment]
Wanted to quickly let y'all on about some [helper scripts](https://github.com/jonico/pscale-cli-helper-scripts) our own @jonico put together.  Check out the section on [GitHub Action integrations](https://github.com/jonico/pscale-cli-helper-scripts#zero-setup-action-workflows-included-sparkles).
